docs/api/isAllOf.mdx

```diff
@@ -0,0 +1,180 @@
+  return action.payload.isInteresting;
+}
+
+interface ExampleState {
+  isSpecial: boolean;
+  isInteresting: boolean;
+}
+
+const initialState: ExampleState = {
+  isSpecial: false,
+  isInteresting: false,
+}
+
+const request = createAsyncThunk('request', () => {
+  return {
+    isSpecial: true,
+    isInteresting: true,
+  };
+});
+
+const loadingReducer = createReducer(initialState, builder => {
+  builder
+    .addCase(request.fulfilled, (state, action) => {
+      if (isSpecial(action)) {
+        return {
+          ...state,
+          isSpecial: true,
+        }
+      } else if (isInteresting(action)) {
+        return {
+          ...state,
+          isInteresting: true,
+        }
+      }
+    });
+});
+```
+
+With isAllOf, we can use matchers to reorganize that code into a more canonical form.
+And more importantly, it enables more complex cases to be handled straightforwardly.
+
+```ts
+import { createAsyncThunk, createReducer, PayloadAction, isAllOf } from '@reduxjs/toolkit'
+
+interface Data {
+  isInteresting: boolean;
+  isSpecial: boolean;
+}
+
+interface Special extends Data {
+  isSpecial: true;
+}
+
+interface Interesting extends Data {
+  isInteresting: true;
+}
+
+function isSpecial(action: PayloadAction<Data>): action is PayloadAction<Special> {
+  return action.payload.isSpecial;
+}
+
+function isInteresting(action: PayloadAction<Data>): action is PayloadAction<Interesting> {
+  return action.payload.isInteresting;
+}
+
+interface ExampleState {
+  isSpecial: boolean;
+  isInteresting: boolean;
+}
+
+const initialState: ExampleState = {
+  isSpecial: false,
+  isInteresting: false,
+}
+
+const request = createAsyncThunk('request', () => {
+  return {
+    isSpecial: true,
+    isInteresting: true,
+  };
+});
+
+const loadingReducer = createReducer(initialState, builder => {
+  builder
+    .addMatcher(isAllOf(request.fulfilled, isSpecial), (state, action) => {
+      return {
+        ...state,
+        isSpecial: true,
+      }
+    })
+    .addMatcher(isAllOf(request.fulfilled, isInteresting), (state, action) => {
+      return {
+        ...state,
+        isInteresting: true,
+      }
+    });
+})
+```
+
+## As a TypeScript TypeGuard
+
+The function `isAllOf` also accepts regular TypeScript type guards, and the function returned
+by `isAllOf` can also be used as a TypeScript type guard in other contexts.
+
+```ts
+import { isAllOf, PayloadAction } from '@reduxjs/toolkit'
+
+interface Data {
+  isInteresting: boolean;
+  isSpecial: boolean;
+}
+
+interface Special extends Data {
+  isSpecial: true;
+}
+
+interface Interesting extends Data {
+  isInteresting: true;
+}
+
+function isSpecial(action: PayloadAction<Data>): action is PayloadAction<Special> {
+  return action.payload.isSpecial;
+}
+
+function isInteresting(action: PayloadAction<Data>): action is PayloadAction<Interesting> {
+  return action.payload.isInteresting;
+}
```

Comment:
I think once you merge these into one page, there will be no need to show these same types & type guards this many times. I feel like the bloat takes away from where the focus should be below.